apiserver/facades/agent/uniter/uniter.go

```diff
@@ -1645,7 +1681,18 @@ func (u *UniterAPI) NetworkInfo(args params.NetworkInfoParams) (params.NetworkIn
 		}
 	}
 
-	networkInfos := machine.GetNetworkInfoForSpaces(spaces)
+	// Don't get the network info for the default space if there was an error
+	// getting the address.
+	var defaultSpaceNetworkErr error
+	if spaceAddressErr != nil && spaces.Contains(environs.DefaultSpaceName) {
+		spaces.Remove(environs.DefaultSpaceName)
+		defaultSpaceNetworkErr = errors.Annotatef(err, "getting machine %q preferred primary address", machine.MachineTag())
+	}
+
+	networkInfos := machine.GetNetworkInfoForSpaces(spaces, defaultSpaceAddress)
```

Comment:
A concrete case I have seen is a charm calling "unit-get private-address". But network-get also needs to be relation aware as well. The address for the default space does then become contextually dependent on whether it's a cross model relation, so long as the relation endpoint is bound to that space (I think that's the issue here? - we need to consider the endpoint binding and only if the binding is omitted or explicitly specifies the default space do we need to provide the calculated address?)